File: ios/Shared/StatusMark.swift
// The Remotly mark as a status glyph (shared/design/DESIGN.md §4.12): the icon's chevron in the status colour and the
// teal phone, on the icon's 64-unit grid cropped to x 10–52, y 14–50 (42 × 36 units), scaled to `height`. Compiled
// into the app and the FlowActivity widget extension; shapes only (no Canvas), so Live Activities can draw it.
import SwiftUI

struct StatusMark: View {
    /// Chevron colour = the status colour (§3); `fg3` for a plain shell.
    var chevron: Color
    /// Height in points; the width follows the 42 : 36 crop.
    var height: CGFloat = 14
    /// Shift of the chevron from its resting place, in grid units (toward the phone is positive).
    var chevronOffset: CGFloat = 0
    var chevronOpacity: Double = 1
    /// A second chevron 6 units behind at 35 %: the drawn motion of the Live Activity while working.
    var trail = false

    private var scale: CGFloat { height / 36 }

    var body: some View {
        ZStack(alignment: .topLeading) {
            // Phone: grid x 35, y 14, 17 × 36, radius 5.4; island cut-out x 40.5, y 16.6, 6 × 2.2, radius 1.1.
            RoundedRectangle(cornerRadius: 5.4 * scale, style: .continuous)
                .fill(DesignTokens.accent)
                .frame(width: 17 * scale, height: 36 * scale)
                .offset(x: 25 * scale, y: 0)
            RoundedRectangle(cornerRadius: 1.1 * scale, style: .continuous)
                .fill(DesignTokens.bg)
                .frame(width: 6 * scale, height: 2.2 * scale)
                .offset(x: 30.5 * scale, y: 2.6 * scale)
            if trail {
                ChevronShape()
                    .fill(chevron.opacity(0.35 * chevronOpacity))
                    .frame(width: 17.2 * scale, height: 25 * scale)
                    .offset(x: (0.2 + chevronOffset - 6) * scale, y: 5.5 * scale)
            }
            // Chevron: grid M15.6 19.5 L27.4 32 L15.6 44.5 L10.2 44.5 L22 32 L10.2 19.5 Z.
            ChevronShape()
                .fill(chevron.opacity(chevronOpacity))
                .frame(width: 17.2 * scale, height: 25 * scale)
                .offset(x: (0.2 + chevronOffset) * scale, y: 5.5 * scale)
        }
        .frame(width: 42 * scale, height: 36 * scale, alignment: .topLeading)
        .clipped() // shapes carry no accessibility of their own; callers label the mark with the status word
    }
}

/// The icon's chevron in its own 17.2 × 25 box (grid points relative to x 10.2, y 19.5).
struct ChevronShape: Shape {
    func path(in rect: CGRect) -> Path {
        func point(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
            CGPoint(x: rect.minX + x / 17.2 * rect.width, y: rect.minY + y / 25 * rect.height)
        }
        var path = Path()
        path.move(to: point(5.4, 0))
        path.addLine(to: point(17.2, 12.5))
        path.addLine(to: point(5.4, 25))
        path.addLine(to: point(0, 25))
        path.addLine(to: point(11.8, 12.5))
        path.addLine(to: point(0, 0))
        path.closeSubpath()
        return path
    }
}

/// The working loop of §4.12, shared by the app's animated mark: 1.4 s, ease-in-out from x −3 to +7 units, opaque for
/// two thirds of the way and fading to 0 over the last third. `phase` is 0…1 within the loop.
enum StatusMarkMotion {
    static let period: TimeInterval = 1.4

    static func phase(at date: Date) -> Double {
        date.timeIntervalSinceReferenceDate.truncatingRemainder(dividingBy: period) / period
    }

    static func offset(phase t: Double) -> CGFloat {
        let eased = t < 0.5 ? 2 * t * t : 1 - pow(-2 * t + 2, 2) / 2
        return CGFloat(-3 + 10 * eased)
    }

    static func opacity(phase t: Double) -> Double {
        t < 2.0 / 3.0 ? 1 : max(0, 1 - (t - 2.0 / 3.0) * 3)
    }
}

File: android/app/src/main/kotlin/com/inferenceaftermath/remotly/ui/StatusMark.kt
// The Remotly mark as a status glyph (shared/design/DESIGN.md §4.12): the icon's chevron in the status colour and the
// teal phone, on the icon's 64-unit grid cropped to x 10–52, y 14–50 (42 × 36 units), scaled to `height`. Same geometry
// and motion as the iOS `StatusMark` / `PaneMark`.
package com.inferenceaftermath.remotly.ui

import android.database.ContentObserver
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.clipRect
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.inferenceaftermath.remotly.core.protocol.AgentStatus
import com.inferenceaftermath.remotly.core.protocol.PaneInfo
import kotlin.math.pow

/**
 * The mark: phone always `accent` with the `bg` island; the chevron in [chevron] (the status colour), shifted by
 * [chevronOffset] grid units toward the phone, at [chevronOpacity]; [trail] draws a second chevron 6 units behind at
 * 35 % (the notification / drawn-motion variant).
 */
@Composable
fun StatusMark(
    chevron: Color,
    height: Dp = 14.dp,
    chevronOffset: Float = 0f,
    chevronOpacity: Float = 1f,
    trail: Boolean = false,
    modifier: Modifier = Modifier,
) {
    Canvas(modifier.size(height * 42f / 36f, height)) {
        val s = size.height / 36f
        fun x(u: Float) = (u - 10f) * s
        fun y(u: Float) = (u - 14f) * s
        drawRoundRect(Tokens.accent, topLeft = Offset(x(35f), y(14f)), size = Size(17f * s, 36f * s), cornerRadius = CornerRadius(5.4f * s))
        drawRoundRect(Tokens.bg, topLeft = Offset(x(40.5f), y(16.6f)), size = Size(6f * s, 2.2f * s), cornerRadius = CornerRadius(1.1f * s))
        fun chevronPath(dx: Float) = Path().apply {
            moveTo(x(15.6f + dx), y(19.5f)); lineTo(x(27.4f + dx), y(32f)); lineTo(x(15.6f + dx), y(44.5f))
            lineTo(x(10.2f + dx), y(44.5f)); lineTo(x(22f + dx), y(32f)); lineTo(x(10.2f + dx), y(19.5f)); close()
        }
        clipRect {
            if (trail) drawPath(chevronPath(chevronOffset - 6f), chevron.copy(alpha = chevron.alpha * 0.35f * chevronOpacity))
            drawPath(chevronPath(chevronOffset), chevron.copy(alpha = chevron.alpha * chevronOpacity))
        }
    }
}

/**
 * A pane's mark: the chevron in the pane's status colour; while the pane is working and [animated] is on it glides
 * toward the phone in the 1.4 s loop of §4.12 (rows stay still, the pane header animates). Still when the system
 * removes animations. Labelled with the status word ("Terminal" for a plain shell).
 */
@Composable
fun PaneMark(p: PaneInfo, height: Dp = 14.dp, animated: Boolean = false, modifier: Modifier = Modifier) {
    val color = Status.color(p)
    // The status word; an agent whose status has no word (unknown) is named; "Terminal" only for a plain shell.
    val label = Status.word(p) ?: if (p.hasAgent) (agentName(p) ?: "Agent") else "Terminal"
    val reduceMotion = if (animated) rememberReduceMotion() else false
    val moving = animated && !reduceMotion && p.hasAgent && p.agent_status == AgentStatus.WORKING
    val labelled = modifier.semantics { contentDescription = label }
    if (moving) {
        val phase by rememberInfiniteTransition(label = "mark").animateFloat(
            initialValue = 0f, targetValue = 1f,
            animationSpec = infiniteRepeatable(tween(durationMillis = 1400, easing = LinearEasing)),
            label = "phase",
        )
        StatusMark(color, height, chevronOffset = markOffset(phase), chevronOpacity = markOpacity(phase), modifier = labelled)
    } else {
        StatusMark(color, height, modifier = labelled)
    }
}

/**
 * Whether the system removes animations (Settings › Accessibility › "Remove animations" sets the animator scale to 0),
 * observed live so a mark on screen goes still the moment the setting changes.
 */
@Composable
private fun rememberReduceMotion(): Boolean {
    val resolver = LocalContext.current.contentResolver
    fun read() = Settings.Global.getFloat(resolver, Settings.Global.ANIMATOR_DURATION_SCALE, 1f) == 0f
    var reduced by remember(resolver) { mutableStateOf(read()) }
    DisposableEffect(resolver) {
        val observer = object : ContentObserver(Handler(Looper.getMainLooper())) {
            override fun onChange(selfChange: Boolean) { reduced = read() }
        }
        resolver.registerContentObserver(Settings.Global.getUriFor(Settings.Global.ANIMATOR_DURATION_SCALE), false, observer)
        onDispose { resolver.unregisterContentObserver(observer) }
    }
    return reduced
}

/** Ease-in-out from x −3 to +7 grid units over the loop (`phase` 0…1). */
fun markOffset(phase: Float): Float {
    val eased = if (phase < 0.5f) 2f * phase * phase else 1f - (-2f * phase + 2f).pow(2) / 2f
    return -3f + 10f * eased
}

/** Opaque for two thirds of the loop, then fading to 0. */
fun markOpacity(phase: Float): Float = if (phase < 2f / 3f) 1f else (1f - (phase - 2f / 3f) * 3f).coerceAtLeast(0f)
